GitExtensions/Project.Publish.targets

```diff
@@ -0,0 +1,84 @@
+<Project>
+
+  <!--
+    ============================================================
+                       _PublishTranslations
+
+    Copies all available translations to the publish directory.
+    ============================================================
+    -->
+  <Target Name="_PublishTranslations">
+    <Copy SourceFiles="@(Translations)" DestinationFolder="$(PublishDir)Translation" ContinueOnError="false" />
+  </Target>
+
+  <!--
+    ============================================================
+                       _PublishExtraDependecies
+
+    Copies additional files (e.g. EasyHook dlls) to the publish directory.
+    ============================================================
+    -->
+  <Target Name="_PublishExtraDependecies">
+    <Copy SourceFiles="$(SolutionDir)\Externals\EasyHook\EasyHookDll\Build\Release\x64\EasyHook64.dll" DestinationFolder="$(PublishDir)" />
+    <Copy SourceFiles="$(SolutionDir)\Externals\EasyHook\EasyHookDll\Build\Release\x86\EasyHook32.dll" DestinationFolder="$(PublishDir)" />
+  </Target>
+
+  <!--
+    ============================================================
+                       _RetrieveCurrentBuildVersion
+
+    Retrieves the current build version.
+    ============================================================
+    -->
+  <Target Name="_RetrieveCurrentBuildVersion">
+    <PropertyGroup>
+      <ExeAssembly>$([MSBuild]::NormalizePath('$(PublishDir)', $(_DeploymentApplicationManifestIdentity)))</ExeAssembly>
+    </PropertyGroup>
+
+    <GetAssemblyIdentity AssemblyFiles="$(ExeAssembly)">
+      <Output TaskParameter="Assemblies" ItemName="ExeAssemblyIdentities"/>
+    </GetAssemblyIdentity>
+
+    <PropertyGroup>
+      <CurrentBuildVersion>%(ExeAssemblyIdentities.Version)</CurrentBuildVersion>
+    </PropertyGroup>    
+  </Target>
+
+  <!--
+    ============================================================
+                       CreatePortable
+
+    Creates a portable archive.
+    ============================================================
+    -->
+
+  <Target Name="CreatePortable" AfterTargets="Publish" DependsOnTargets="_RetrieveCurrentBuildVersion;_PublishTranslations;_PublishExtraDependecies">
+    <PropertyGroup>
+      <!-- Resolve app.config, so we can set/unset "portable" flag -->
+      <_PublishAppConfig>$([System.IO.Path]::GetFileName('$(TargetAppConfig)'))</_PublishAppConfig>
+      <_PublishAppConfigPath>$([System.IO.Path]::Combine('$(PublishDir)', '$(_PublishAppConfig)'))</_PublishAppConfigPath>
+
+      <!-- Resolve the output file -->
+      <_PublishPortableFileName>Portable-$(CurrentBuildVersion.ToString()).zip</_PublishPortableFileName>
```

Comment:
Is it possible to add 'GitExt' or 'GitExtensions' at the beginning of the zip file.
I think that will make it more identifiable in a full download folder ;)

And also **if it's really straight forward** to do, add the hash at the end by following the SemVer syntax (i.e  `GitExtensions-Portable-4.0.0.8598+f0d22742.zip`)

Reply:
Yes, it is easy to add a prefix:
```xml
<_PublishPortableFileName>GitExtensions-Portable-$(CurrentBuildVersion.ToString()).zip</_PublishPortableFileName>
```

A hash may be a (little?) more involved, but doable.

Reply:
Done. 
For non-release builds I propose we just have a commit hash instead of a build number, e.g. **GitExtensions-portable-f511a73a7.zip**. For a release we'll have a build number, e.g. **GitExtensions-portable-3.3.0.2342.zip**.
What do you think?

Reply:
I think that's good for now.

It will be perfect for 'master' builds and releases. No so much for PR (I thought about it afterwards...) 

I will try to look **later** at how to  retrieve the hash of the head of the pull request and use it as the hash here.

Because Appveyor build a commit which is a merge of the PR within 'master' so the hash will be of no use in this case 😕

Reply:
https://www.appveyor.com/docs/environment-variables/

* APPVEYOR_PULL_REQUEST_HEAD_COMMIT - Pull (Merge) Request source commit ID (SHA)

Or more likely :

* APPVEYOR_REPO_COMMIT - commit ID (SHA)

I think that we should add something like that to override the value on Appveyor (in pseudo msbuild on smartphone 😁) :

`<_PublishPortableFileNameSuffix Condition="'$(GitCommit)' != '' And '$(env:APPVEYOR_REPO_COMMIT)' !='' ">-$(env:APPVEYOR_REPO_COMMIT)</_PublishPortableFileNameSuffix>`

Reply:
> For non-release builds I propose we just have a commit hash instead of a build number, e.g. GitExtensions-portable-f511a73a7.zip. For a release we'll have a build number, e.g. GitExtensions-portable-3.3.0.2342.zip.

Build numbers have the advantage that they can be used for a kind of bisecting without building the project. I'd prefer to have it with build number and with hash appended.

> dirty

I agree only if AppVeyor builds a temporary merge with master.
But I don't like this for two reasons:
1. AppVeyor denies to build if there are conflicts with master. So you cannot finish the feature without being forced to merge the master branch.
2. The resulting build is a moving target depending on the master branch.
I'd prefer that AppVeyor builds exactly the _clean_ commit for PRs.

Reply:
> Build numbers have the advantage that they can be used for a kind of bisecting without building the project. I'd prefer to have it with build number and with hash appended.

Apologies, I don't follow the logic here. How is a build number useful for bisecting, e.g. 4.0.0.1234? I can see a hash being useful though.



> I'd prefer that AppVeyor builds exactly the _clean_ commit for PRs.

The current function of "dirty" flag is to indicate whether a build is an official release or not. It is useful when users send us bug reports.
If we mark all builds as "clean" we won't know whether a specific build is official or not _without going to the GitHub release tab and checking there_.

> 1. AppVeyor denies to build if there are conflicts with master

If there is a conflict, it has to be resolved one way or another. I don't follow how it relates to the "dirty" flag discussion, sorry.

Reply:
> bisecting

Usually, I have many downloaded portable builds. They are well-ordered by build numbers.
Checking different builds for an issue with commits hashs only, would be a pain.

> dirty

My understanding of "dirty" is "There are uncommitted changes."
If I really forgot the build number of a release, I will see it in my folder with the portable builds (last one).

> If there is a conflict, it has to be resolved one way or another.

Yes, but we should allow to finish the implementation and being able to check it by AppVeyor before it is merged with master.

> I don't follow how it relates to the "dirty" flag discussion, sorry.

If AppVeyor builds a temporary merge with master, this _is_ a "dirty" build because it is no official commit.

Reply:
> this _is_ a "dirty" build because it is no official commit.

You mis-understand, 'clean' builds only come from release branches. All other commits are marked dirty. This is how we had it since we introduced GitInfo into our builds.

A 'dirty' build means there are uncommitted changes in the working directory. We inject build numbers into AssemblyInfo and C++ header files, and that makes builds 'dirty'. For release builds (non-pr commits on 'release' branches) we ignore these changes making builds 'clean'.


> Usually, I have many downloaded portable builds. They are well-ordered by build numbers.
> Checking different builds for an issue with commits hashs only, would be a pain.

Make sense.

Reply:
I don't mis-understand this IMO. I just had not in mind, that a non-release build is modified in this way and hence is really a "dirty" build from git point of view. Though IMO this is just a technical issue how we inject build numbers. Whether this makes a commit build dirty or not, is a non-productive philosophical question. I can live with "dirty" non-release builds.

Reply:
> I agree only if AppVeyor builds a temporary merge with master.
>[...] 
>I'd prefer that AppVeyor builds exactly the clean commit for PRs.

I think that is GitHub that trigger Appveyor builds only if there are no conflicts.
And also GitHub trigger it 2 times for each PR. One is to build only the PR. The other is for the PRs merged with 'master' (to do real Continuous Integration).

That makes me think of something important. What is the build result that we consider to be the 'dev' version and that we should propose to download and test.

I think that it's better to use the 'merged' one.

Perhaps also we should do something to differentiate the artifacts because they will have the same name 🤔

Reply:
> And also GitHub trigger it 2 times for each PR. One is to build only the PR. The other is for the PRs merged with 'master' (to do real Continuous Integration).

This only happens for commits on the main trunk. It doesn't happen for PRs coming from forks.

Reply:
Thanks for the clarification.